### PR TITLE
[drape] Improved animation in follow-and-rotate mode

### PR DESCRIPTION
**Pull request by rokuz**

### CODE REVIEW

**Comment by mapsmetest** [comment]
Jenkins asks: can anyone have a look at this request?
If you want Jenkins to compile and test this, write "test this please" or JT_TP (without the underscore in the middle).

**Comment by ExMix** [comment]
Выглядит красиво

File: drape_frontend/animation/model_view_animation.cpp [hunk 1]
```diff
@@ -91,34 +91,38 @@ double ModelViewAnimation::GetScaleDuration(double startSize, double endSize)
...
 {
   ScreenBase s = screen;
   s.SetFromRect(rect);
-  m2::PointD const p = s.PtoG(m_pixelPoint);
-  rect.Offset(m_globalPoint - p);
+
+  m2::PointD const pixelPoint = s.GtoP(m_globalPoint);
+  m2::PointD const newCenter = s.PtoG(pixelPoint + m_pixelOffset);
+
+
```
Comment (deathbaba) [inline drape_frontend/animation/model_view_animation.cpp]: extra \n ?

File: drape_frontend/my_position_controller.cpp [hunk 2]
```diff
@@ -178,6 +178,12 @@ void MyPositionController::CorrectScalePoint(m2::PointD & pt1, m2::PointD & pt2)
   }
 }
 
+void MyPositionController::CorrectGlobalScalePoint(m2::PointD & pt) const
```
Comment (deathbaba) [inline drape_frontend/my_position_controller.cpp]: Confusing name for the method. Would it be easier to return point and name it like GetXXX?

File: drape_frontend/navigator.cpp [hunk 3]
```diff
@@ -407,13 +412,13 @@ bool Navigator::CheckBorders(ScreenBase const & screen) const
 
 bool Navigator::ScaleImpl(m2::PointD const & newPt1, m2::PointD const & newPt2,
                           m2::PointD const & oldPt1, m2::PointD const & oldPt2,
-                          bool skipMinScaleAndBordersCheck,
-                          bool doRotateScreen)
+                          bool skipMinScaleAndBordersCheck, bool doRotateScreen,
+                          ScreenBase & screen)
 {
-  math::Matrix<double, 3, 3> newM = m_Screen.GtoPMatrix() * ScreenBase::CalcTransform(oldPt1, oldPt2, newPt1, newPt2);
+  math::Matrix<double, 3, 3> newM = screen.GtoPMatrix() * ScreenBase::CalcTransform(oldPt1, oldPt2, newPt1, newPt2);
```
Comment (deathbaba) [inline drape_frontend/navigator.cpp]: const?

File: drape_frontend/user_event_stream.cpp [hunk 4]
```diff
@@ -217,24 +220,25 @@ bool UserEventStream::SetScale(m2::PointD const & pxScaleCenter, double factor,
...
 
-    m2::RectD sizeRect = rect.GetLocalRect();
-    sizeRect.Scale(1.0 / factor);
+    ScreenBase screen = GetCurrentScreen();
+    m_navigator.CalculateScale(scaleCenter, factor, screen);
+    m2::PointD offset = GetCurrentScreen().PixelRect().Center() - scaleCenter;
 
-    auto creator = [this, &scaleCenter, &glbScaleCenter](m2::AnyRectD const & startRect, m2::AnyRectD const & endRect,
-                                                         double aDuration, double mDuration, double sDuration)
+    auto creator = [this, &glbScaleCenter, &offset](m2::AnyRectD const & startRect, m2::AnyRectD const & endRect,
```
Comment (deathbaba) [inline drape_frontend/user_event_stream.cpp]: const?

File: drape_frontend/user_event_stream.cpp [hunk 5]
```diff
@@ -314,6 +324,16 @@ bool UserEventStream::SetFollowAndRotate(m2::AnyRectD const & rect, m2::PointD c
   return true;
 }
 
+void UserEventStream::ResetCurrentAnimation()
+{
+  if (m_animation != nullptr)
```
Comment (deathbaba) [inline drape_frontend/user_event_stream.cpp]: if (m_animation) ?

File: drape_frontend/user_event_stream.cpp [hunk 6]
```diff
@@ -314,6 +324,16 @@ bool UserEventStream::SetFollowAndRotate(m2::AnyRectD const & rect, m2::PointD c
...
+  {
+    m2::AnyRectD rect = m_animation->GetCurrentRect(GetCurrentScreen());
```
Comment (deathbaba) [inline drape_frontend/user_event_stream.cpp]: const?

**Comment by deathbaba** [comment]
LGTM with minor comments.

File: drape_frontend/my_position_controller.cpp (hunk 2 above)
Comment (rokuz) [inline drape_frontend/my_position_controller.cpp]: Семантически будет неверно, метод должен корректирвать переданную точку

File: drape_frontend/animation/model_view_animation.cpp [hunk 7]
```diff
@@ -91,68 +91,96 @@ double ModelViewAnimation::GetScaleDuration(double startSize, double endSize)
...
+  return CalculateCenter(screen.GlobalRect().GetLocalRect(), screen.PixelRect(), userPos, pixelPos, azimuth);
 }
 
-m2::AnyRectD FollowAndRotateAnimation::GetRect(double elapsedTime) const
+m2::PointD FollowAndRotateAnimation::CalculateCenter(m2::RectD const & localRect, m2::RectD const & pixelRect,
+                                                     m2::PointD const & userPos, m2::PointD const & pixelPos,
+                                                     double azimuth)
+{
+  m2::PointD formingVector = pixelRect.Center() - pixelPos;
+  formingVector.x /= pixelRect.SizeX();
```
Comment (darina) [inline drape_frontend/animation/model_view_animation.cpp]: Можно немного упростить
formingVector.x *= localRect.SizeX() / pixelRect.SizeX();

**Comment by darina** [comment]
LGTM

**Pull request merged**